src/index.js

```diff
@@ -19,177 +19,264 @@ const debug = (function () {
 })()
 
 class Masq {
-  /**
-   * constructor
-   * @param {string} appName - The application name
-   */
   constructor (appName, appDescription, appImageURL, options = {}) {
-    this._reset()
-
     this.eventTarget = document.createElement('MasqLib')
 
     this.appName = appName
     this.appDescription = appDescription
     this.appImageURL = appImageURL
 
-    this.userId = null
-    this.dataEncryptionKey = null
-    this.userAppDb = null
-
-    this._loadSessionInfo()
-
-    // login state data
-    this.loginChannel = null
-    this.loginKey = null
-    this.loginUrl = null
-
     // override config with constructor options
     this.config = {
       hubUrls: options.hubUrls ? options.hubUrls : jsonConfig.hubUrls,
       masqAppBaseUrl: options.masqAppBaseUrl ? options.masqAppBaseUrl : jsonConfig.masqAppBaseUrl,
       swarmConfig: options.swarmConfig ? options.swarmConfig : jsonConfig.swarmConfig
     }
 
+    // init state
+    this.state = 'notLogged'
+
+    // login state data
+    this.stayConnected = false
+    this.loginChannel = null
+    this.loginKey = null
+    this.loginSw = null
+    this.loginPeer = null
+
+    // reset replication variables
+    this.replicating = false
+    this.userAppRepSW = null
+    this.userAppRepHub = null
+
+    // setup listeners
     this._listenForLoginOrSignout()
     this._listenForFocus()
+
+    // init state through async function
+    // this.init should be awaited to make sure state is initialized
+    this.initPromise = this._init()
+  }
+
+  async setState (newState) {
+    const currState = this.state
+
+    // BEFORE
+    switch (newState) {
+      case 'notLogged':
+        break
+      case 'userAppDbCreated':
+        break
+      case 'accessMaterialReceived':
+        break
+      case 'registerNeeded':
+        break
+      case 'authorized':
+        break
+      case 'loggingIn':
```

Comment:
Can we remove all the empty cases ?

Reply:
Yes we could, or at least comment them. I wrote them to list states.